src/main/kotlin/com/github/gradle/node/NodeExtension.kt

```diff
@@ -1,38 +1,57 @@
 package com.github.gradle.node
 
-import com.github.gradle.node.variant.Variant
 import org.gradle.api.Project
-import java.io.File
-import kotlin.properties.Delegates
+import org.gradle.kotlin.dsl.create
+import org.gradle.kotlin.dsl.getByType
+import org.gradle.kotlin.dsl.property
 
 open class NodeExtension(project: Project) {
-    private val cacheDir = File(project.projectDir, ".gradle")
-    var workDir = File(cacheDir, "nodejs")
-    var npmWorkDir = File(cacheDir, "npm")
-    var yarnWorkDir = File(cacheDir, "yarn")
-    var nodeModulesDir: File = project.projectDir
-    var version = "12.16.1"
-    var npmVersion = ""
-    var yarnVersion = ""
-    var distBaseUrl: String? = "https://nodejs.org/dist"
-    var npmCommand = "npm"
-    var npxCommand = "npx"
-    var npmInstallCommand = "install"
-    var yarnCommand = "yarn"
-    var download = false
-    var variant by Delegates.notNull<Variant>()
+    private val cacheDir = project.layout.projectDirectory.dir(".gradle")
+    val workDir = project.objects.directoryProperty().convention(cacheDir.dir("nodejs"))
+    val npmWorkDir = project.objects.directoryProperty().convention(cacheDir.dir("npm"))
+    val yarnWorkDir = project.objects.directoryProperty().convention(cacheDir.dir("yarn"))
+    val nodeModulesDir = project.objects.directoryProperty().convention(project.layout.projectDirectory)
+    val version = project.objects.property<String>().convention("12.16.2")
+    val npmVersion = project.objects.property<String>().convention("")
+    val yarnVersion = project.objects.property<String>().convention("")
+    val distBaseUrl = project.objects.property<String>()
+    val npmCommand = project.objects.property<String>().convention("npm")
+    val npxCommand = project.objects.property<String>().convention("npx")
+    val npmInstallCommand = project.objects.property<String>().convention("install")
+    val yarnCommand = project.objects.property<String>().convention("yarn")
+    val download = project.objects.property<Boolean>().convention(false)
+
+    init {
+        distBaseUrl.set("https://nodejs.org/dist")
```

Comment:
Does this interfere with `distBaseUrl = null`, the way we disabled automatic adding of repositories (for those with corporate environments)?

Reply:
Yes, I has to do with unsetting this property.

I learned a few important details regarding the `Property` API:

- By default, a property is not defined (its value is null). If you want to unset it, you have to set null value.
- Using a convention is not the same thing as setting an initial value. The convention value is a value that will be used if the value is not defined when retrieving it. That means that if you set a value and unset it, it will return the convention value.
- When the property is not defined (its value is null), calling `.map()` will do nothing. It took me a lot of time to understand that one test was broken because of that. It was about an `@Output` annotation returning a `Provider` (it sounds like it is recommended to do that). I wanted to map an optional configuration value and define the output according to is value. When the property was not set, the `Provider` was considered as empty and the mapping function that handled the `null` case was not called. I had to break the provider chain at this place. I did not find any alternative solution.

To come back to your question, if I use a convention, the property not set but you just have a default value. Unsetting the value won't change anything (since it is already not set). To get it unsettable, we cannot use convention but we have to initially set the value: this will enable the value to be unsettable later.